## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns.tolist()

['first',
 'last',
 'job',
 'phone',
 'email',
 'client_id',
 'order_id',
 'order_date',
 'order_week',
 'order_year',
 'item_id',
 'category',
 'subcategory',
 'unit_price',
 'unit_cost',
 'unit_weight',
 'qty',
 'line_number']

In [3]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.shape
df.dtypes
df.count()
df.isnull()


first   last    job  phone  email  client_id  order_id  order_date  \
0      False  False  False  False  False      False     False       False   
1      False  False  False  False  False      False     False       False   
2      False  False  False  False  False      False     False       False   
3      False  False  False  False  False      False     False       False   
4      False  False  False  False  False      False     False       False   
...      ...    ...    ...    ...    ...        ...       ...         ...   
54634  False  False  False  False  False      False     False       False   
54635  False  False  False  False  False      False     False       False   
54636  False  False  False  False  False      False     False       False   
54637  False  False  False  False  False      False     False       False   
54638  False  False  False  False  False      False     False       False   

       order_week  order_year  item_id  category  subcategory  unit_price  \
0           False       False    False     False        False       False   
1           False       False    False     False        False       False   
2           False       False    False     False        False       False   
3           False       False    False     False        False       False   
4           False       False    False     False        False       False   
...           ...         ...      ...       ...          ...         ...   
54634       False       False    False     False        False       False   
54635       False       False    False     False        False       False   
54636       False       False    False     False        False       False   
54637       False       False    False     False        False       False   
54638       False       False    False     False        False       False   

       unit_cost  unit_weight    qty  line_number  
0          False        False  False        False  
1          False        False  False        False  
2          False        False  False        False  
3          False        False  False        False  
4          False        False  False        False  
...          ...          ...    ...          ...  
54634      False        False  False        False  
54635      False        False  False        False  
54636      False        False  False        False  
54637      False        False  False        False  
54638      False        False  False        False  

[54639 rows x 18 columns]

In [5]:
# What three item categories had the most entries?
df["category"].value_counts()

category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64

In [6]:
# For the category with the most entries, which subcategory had the most entries?
df[df["category"] == "consumables"]["subcategory"].value_counts()


subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
misc                 2891
pens                 2603
Name: count, dtype: int64

In [7]:
# Which five clients had the most entries in the data?
df["client_id"].value_counts()

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
        ... 
70734      4
32001      4
77583      3
76017      3
94987      1
Name: count, Length: 999, dtype: int64

In [8]:
# Store the client ids of those top 5 clients in a list.
top_5_clients_df = df["client_id"].value_counts().head(5).index
top_5_clients_df


Index([33615, 66037, 46820, 38378, 24741], dtype='int64', name='client_id')

In [9]:
# How many total units (the qty column) did the client with the most entries order?
df[df["client_id"] == 33615]["qty"].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["sub_total"] = df["qty"] * df["unit_price"]
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  sub_total  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [11]:
# Create a column for shipping price.
df["total_weight"] = (df["unit_weight"] * df["qty"]).round(2)
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df["shipping_price"] = df.apply(lambda row: row["total_weight"] * 7.00 if row["total_weight"] > 50.00 else row["total_weight"] * 10.00, axis=1)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number  sub_total  total_weight  \
0     762.71         7.50  105            1  115164.00        787.50   
1      15.09         1.49   21            0     523.95         31.29   
2       7.86         1.68   39            6     527.28         65.52   
3      24.85         1.23   29            3    1056.18         35.67   
4     108.17        46.43   20            1    3902.00        928.60   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  

[5 rows x 21 columns]

In [12]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df["tax"] = ((df["sub_total"] + df["shipping_price"]) * .0925).round(2)
df["total_price"] = (df["sub_total"] + df["shipping_price"] + df["tax"]).round(2)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  sub_total  total_weight  shipping_price       tax  total_price  
0            1  115164.00        787.50         5512.50  11162.58    131839.08  
1            0     523.95         31.29          312.90     77.41       914.26  
2            6     527.28         65.52          458.64     91.20      1077.12  
3            3    1056.18         35.67          356.70    130.69      1543.57  
4            1    3902.00        928.60         6500.20    962.20     11364.40  

[5 rows x 23 columns]

In [13]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["total_cost"] = ((df["unit_cost"] * df["qty"]) + df["shipping_price"]).round(2)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   sub_total  total_weight  shipping_price       tax  total_price  total_cost  
0  115164.00        787.50         5512.50  11162.58    131839.08    85597.05  
1     523.95         31.29          312.90     77.41       914.26      629.79  
2     527.28         65.52          458.64     91.20      1077.12      765.18  
3    1056.18         35.67          356.70    130.69      1543.57     1077.35  
4    3902.00        928.60         6500.20    962.20     11364.40     8663.60  

[5 rows x 24 columns]

In [14]:
# Create a column for the profit of each line using line cost and line price
df["profit"] = ((df["total_price"] - df["total_cost"]))
#df_sorted = df.sort_values(by="profit", ascending=False)
#df_sorted.head()
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_weight  qty line_number  sub_total  \
0          17        2023  ...        7.50  105           1  115164.00   
1          20        2023  ...        1.49   21           0     523.95   
2           4        2023  ...        1.68   39           6     527.28   
3          17        2023  ...        1.23   29           3    1056.18   
4          21        2023  ...       46.43   20           1    3902.00   

   total_weight  shipping_price       tax  total_price  total_cost    profit  
0        787.50         5512.50  11162.58    131839.08    85597.05  46242.03  
1         31.29          312.90     77.41       914.26      629.79    284.47  
2         65.52          458.64     91.20      1077.12      765.18    311.94  
3         35.67          356.70    130.69      1543.57     1077.35    466.22  
4        928.60         6500.20    962.20     11364.40     8663.60   2700.80  

[5 rows x 25 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above
order_id_2742071_df = df[df["order_id"] == 2742071]
print(f"The total price for order ID 2742071 is: ${order_id_2742071_df['total_price'].sum():,.2f}")

order_id_2173913_df = df[df["order_id"] == 2173913]
print(f"The total price for order ID 2173913 is: ${order_id_2173913_df['total_price'].sum():,.2f}")

order_id_6128929_df = df[df["order_id"] == 6128929]
print(f"The total price for order ID 6128929 is: ${order_id_6128929_df['total_price'].sum():,.2f}")



The total price for order ID 2742071 is: $152,811.89
The total price for order ID 2173913 is: $162,388.71
The total price for order ID 6128929 is: $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
#sorted_top_5_clients_qty = df.sort_values(by='qty', ascending=False)
#top_5_clients = sorted_top_5_clients_qty['client_id'].head(5).unique()

for client_id in top_5_clients_df:
    total_amount_spent = df[df['client_id'] == client_id]['total_price'].sum()
    print(f"Client {client_id} spent a total of ${total_amount_spent:,.2f}.")


Client 33615 spent a total of $8,377,308.52.
Client 66037 spent a total of $10,259,514.79.
Client 46820 spent a total of $9,743,794.36.
Client 38378 spent a total of $12,906,550.87.
Client 24741 spent a total of $82,268,892.02.


In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_5_client_data = [
    {
        'client_id': client_id,
        'total_units': df[df['client_id'] == client_id]['qty'].sum(),
        'total_shipping': df[df['client_id'] == client_id]['shipping_price'].sum(),
        'total_amount_spent': df[df['client_id'] == client_id]['total_price'].sum(),
        'total_profit': df[df['client_id'] == client_id]['profit'].sum()
    }
    for client_id in top_5_clients_df
]

top_5_clients_details = pd.DataFrame(top_5_client_data)
top_5_clients_details

client_id  total_units  total_shipping  total_amount_spent  total_profit
0      33615        64313      1828984.89          8377308.52    2201994.61
1      66037        43018      1395151.85         10259514.79    3255031.81
2      46820        75768      1601448.84          9743794.36    2736602.72
3      38378        73667      3429455.40         12906550.87    3271829.89
4      24741       239862      5126448.37         82268892.02   36579992.31

In [18]:
# Format the data and rename the columns to names suitable for presentation.
top_5_clients_details.rename(columns={"client_id" : "Client ID",
                                      "total_units" : "Total Units",
                                      "total_shipping": "Total Shipping Cost",
                                      "total_amount_spent" : "Total Spent",
                                      "total_profit" : "Total Profit"}, inplace=True)

# Define the money columns. 
## Total Shipping Cost, Total Spent, and Total Profit are money columns

# Define a function that converts a dollar amount to millions.
def convert_to_millions(currency_value):
    return round(currency_value / 1000000, 2)

# Apply the function to a column in your DataFrame

top_5_clients_details["Total Profit"] = top_5_clients_details["Total Profit"].apply(convert_to_millions)
top_5_clients_details["Total Spent"] = top_5_clients_details["Total Spent"].apply(convert_to_millions)
top_5_clients_details["Total Shipping Cost"] = top_5_clients_details["Total Shipping Cost"].apply(convert_to_millions)

# Rename the columns to reflect the change in the money format. 
top_5_clients_details.rename(columns={"Total Shipping Cost" : "Total Shipping Cost (Millions)",
                                      "Total Spent" : "Total Spent (Millions)",
                                      "Total Profit" : "Total Profit (Millions)"}, inplace=True)

top_5_clients_details

Client ID  Total Units  Total Shipping Cost (Millions)  \
0      33615        64313                            1.83   
1      66037        43018                            1.40   
2      46820        75768                            1.60   
3      38378        73667                            3.43   
4      24741       239862                            5.13   

   Total Spent (Millions)  Total Profit (Millions)  
0                    8.38                     2.20  
1                   10.26                     3.26  
2                    9.74                     2.74  
3                   12.91                     3.27  
4                   82.27                    36.58

In [19]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sort_by_profit_df = top_5_clients_details.sort_values(by='Total Profit (Millions)', ascending=False)

sort_by_profit_df



Client ID  Total Units  Total Shipping Cost (Millions)  \
4      24741       239862                            5.13   
3      38378        73667                            3.43   
1      66037        43018                            1.40   
2      46820        75768                            1.60   
0      33615        64313                            1.83   

   Total Spent (Millions)  Total Profit (Millions)  
4                   82.27                    36.58  
3                   12.91                     3.27  
1                   10.26                     3.26  
2                    9.74                     2.74  
0                    8.38                     2.20